### PR TITLE
overlord/state: introduce cleanup support

### PR DESCRIPTION
**Pull request by niemeyer**

### CODE REVIEW

File: overlord/state/change.go [hunk 1]
```diff
@@ -334,6 +338,27 @@ func (c *Change) taskStatusChanged(t *Task, old, new Status) {
 	c.markReady()
 }
 
+// IsClean returns whether the task has been cleaned. See SetClean.
```
Comment (niemeyer) [inline overlord/state/change.go]: Self review: copy & paste typo here.

File: overlord/state/taskrunner.go [hunk 2]
```diff
@@ -302,6 +326,22 @@ func (r *TaskRunner) Ensure() {
...
 
+func (r *TaskRunner) cleanTask(t *Task) {
+	select {
+	case <-t.Change().Ready():
+	default:
+		return
+	}
+	cleanup, ok := r.cleanups[t.Kind()]
+	if ok {
+		if err := cleanup(t); err != nil {
```
Comment (pedronis) [inline overlord/state/taskrunner.go]: if I understand this is synchronous and done with the State lock held? do we assume they are fast? or do we need to setup goroutines for them like the other handlers, both ways maybe AddCleanup should grow a bit more description about these conditions they are run under

Comment (niemeyer) [inline overlord/state/taskrunner.go]: Good points. Will consider that further and either document or run concurrently and make it typical.

File: overlord/state/change.go (hunk 1 above)
Comment (niemeyer) [inline overlord/state/change.go]: Done.

File: overlord/state/taskrunner.go (hunk 2 above)
Comment (niemeyer) [inline overlord/state/taskrunner.go]: Changed cleanup handlers to be typical, working just like any other handler. This is the most flexible as it won't force timely execution, and the least error prone as we're used to other handlers already.

File: overlord/state/taskrunner.go [hunk 3]
```diff
@@ -177,6 +198,41 @@ func (r *TaskRunner) run(t *Task) {
 	})
 }
 
+func (r *TaskRunner) clean(t *Task) {
+	select {
+	case <-t.Change().Ready():
+	default:
+		return
```
Comment (pedronis) [inline overlord/state/taskrunner.go]: I would add a comment here:

// whole Change is not ready, don't run cleanups yet

**Comment by pedronis** [comment]
small comment, otherwise lgtm

File: overlord/state/taskrunner.go (hunk 3 above)
Comment (niemeyer) [inline overlord/state/taskrunner.go]: Added comment, thanks.

**Pull request merged**